core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -1558,29 +1569,96 @@ private static void rewriteMerge(SqlMerge call) {
             targetTable,
             JoinConditionType.ON.symbol(SqlParserPos.ZERO),
             call.getCondition());
+
+    // Originally, we used the source select of the update statement if it existed.
+    // Now, we always just use select *
+    SqlNodeList topmostSelectList = new SqlNodeList(SqlParserPos.ZERO);
+    topmostSelectList.add(SqlIdentifier.star(SqlParserPos.ZERO));
+
+    // Add all the conditions and values of the matched clauses into the select.
+    // This project is needed to ensure
+    // that nested sub queries are properly evaluated, and any needed joins needed to get the
+    // required subquery values are properly created when expanding the select list
+
+    for (int i = 0; i < call.getMatchedCallList().size(); i++) {
+      SqlNode curMatchCall = call.getMatchedCallList().get(i);
+      if (curMatchCall instanceof SqlUpdate) {
+        SqlUpdate curUpdateCall = (SqlUpdate) curMatchCall;
+        @Nullable SqlNode curCond = curUpdateCall.getCondition();
+        if (curCond != null) {
+          topmostSelectList.add(curCond);
+        } else {
+          topmostSelectList.add(SqlLiteral.createBoolean(true, curUpdateCall.getParserPosition()));
+        }
+        for (int j = 0; j < curUpdateCall.getSourceExpressionList().size(); j++) {
+          SqlNode insertVal = curUpdateCall.getSourceExpressionList().get(j);
+          topmostSelectList.add(insertVal);
+        }
+      } else {
+        SqlDelete curDeleteCall = (SqlDelete) curMatchCall;
+        @Nullable SqlNode curCond = curDeleteCall.getCondition();
+        if (curCond != null) {
+          topmostSelectList.add(curCond);
+        } else {
+          topmostSelectList.add(SqlLiteral.createBoolean(true, curDeleteCall.getParserPosition()));
+        }
+      }
+    }
+
+    // Add all the conditions and values of the Not matched clauses into the select.
+    for (int i = 0; i < call.getNotMatchedCallList().size(); i++) {
```

Comment:
Can we just use `insertCallList` here instead of `call.getNotMatchedCallList()`?